### DISCUSSION TITLE
How to scan using source change date

### DISCUSSION THREAD

**Discussion by ZMcManus** [codefori/vscode-ibmi#951] [Q&A]
Hello,

I a new to VS Code for IBM i. I have used SEU exclusively for IBM i development and VS for some .net development.

I have enabled source dates, but cannot see them in the editor. 
I have tried using the "source dates in gutter" option, but would rather they appear as they do in SEU - at the end of the record. 
Is there a way to make this happen using VS Code for IBM i?

Also, is there a way to search on the source change dates while in an individual source file member?
Often I would like to search for any changes that were done after a specific date.
My preference would be to search line by line - that is, move down (or up) the source file member to the next changed date.

One feature that I like with SEU is the ability to see the number of occurrences of a search string.
Is there a way to do this within VS Code for IBM i?

Can I search and replace on a source change date?
Or reset all selected records to a specific date? This is another feature of SEU that I like to use on occasion.

Thank you,

Zee

**Comment by worksofliam** [comment]
Hi @ZMcManus :) 

> I have tried using the "source dates in gutter" option, but would rather they appear as they do in SEU - at the end of the record. Is there a way to make this happen using VS Code for IBM i?

There is not. We will only support showing source dates in the gutter of the editor.

> One feature that I like with SEU is the ability to see the number of occurrences of a search string. Is there a way to do this within VS Code for IBM i?

Control/Command + F: https://code.visualstudio.com/docs/getstarted/keybindings

> Can I search and replace on a source change date?
> Or reset all selected records to a specific date? This is another feature of SEU that I like to use on occasion.

There is not a way.

> Also, is there a way to search on the source change dates while in an individual source file member?
> Often I would like to search for any changes that were done after a specific date.

In the connection settings, if you enable 'Diff mode' as the source tracker, like this:

<img width="519" alt="image" src="https://user-images.githubusercontent.com/3708366/200849215-89bd37e2-061d-4b81-a2ca-98bc054f5301.png">

After a restart to enable Diff mode, hovering over the gutter will allow you to enter in a source date to filter:

<img width="440" alt="image" src="https://user-images.githubusercontent.com/3708366/200849451-b983f1b7-d96d-40b9-b8f5-50a439c2cf19.png">

Which will then highlight lines that match the filter.

<img width="528" alt="image" src="https://user-images.githubusercontent.com/3708366/200849731-de58d6d7-e78c-4b79-b75b-6dd9e22364a6.png">

**Reply by worksofliam** [reply]
Also, we do not plan on adding more functionality around source dates into VS Code. It likely will not get any more advanced than this for source dates.

**Comment by jkyeung** [comment]
Just wanted to expand on some of Liam's responses:

First of all, I have personally found the "Edit mode" source date tracking nearly unusable. In theory, edit mode ought to be a better analogue to what SEU does, but in practice VS Code doesn't expose enough information (at least not in any reasonably straightforward manner) to the extension to make it workable. Diff mode still has a few quirks, and it works best when (1) your edits are relatively small, and (2) you save your work fairly often. (Actually, saving often is simply one way of helping your edits stay small, since the diff logic compares the current state with the last-saved state, as far as I'm aware. Liam can correct me if I'm wrong about that.)

The source date search in this extension is significantly better than what SEU provides, in two ways: (1) It lets you search for a date range. (2) It highlights all the lines found by your search, so you see them all at once, in context.

Besides hovering over the gutter, you can also access this search by clicking on the "Date Search" area of the status bar on the lower right of the VS Code window. I find this faster and simpler than the hover method.

The number of occurrences of a search string is a standard feature of the find-and-replace functionality in VS Code (and pretty much any other modern editor). It isn't something provided by this extension.

Finally, while adjusting source dates isn't a provided feature, it is relatively easy to do using a bit of SQL. You can enter and execute the SQL with whatever you're comfortable with (e.g. STRSQL, the "Run SQL Scripts" feature in Access Client Solutions, SQL functionality within VS Code provided by either this extension or one of the related ones, etc.). For example, if it has taken me three days to complete one particular modification, I may set all of the relevant lines to a single date (mainly to make it easier for folks searching for it using SEU):

```sql
create alias srcmbr for mylib.qrpglesrc(mymbr);
update srcmbr
    set srcdat = 221109
    where srcdat between 221107 and 221109;
```

---

<strike>P.S. How do you find the number of occurrences of a search string in SEU?</strike> OK, select 2=All for "Occurrences to process" in the Find Options before issuing the search to get a one-time count and reposition you to the first occurrence.

P.P.S. How do you set selected lines (but not all of them) to a specific date in SEU?

**Reply by ZMcManus** [reply]
The easiest way to find all occurrences of a string in SEU is to type   F [ string ] A   on the command line at the top of the editor. More options are available with the F14 key. 
Resetting dates can be accomplished by using the F13 key in SEU - but this will reset ALL dates in the source member. I generally use this to reset all dates to 0 when working with new members that I have copied. I also have a custom PDM option that allows more control over which lines I change a date on - this is generally used when I accidentally change a line.
Thanks to you an Liam for your responses - I am sure that I will have more questions the more I use VS Code for IBM i,
Zee

**Reply by jkyeung** [reply]
> I also have a custom PDM option that allows more control over which lines I change a date on - this is generally used when I accidentally change a line.

Well, in this respect, the diff mode source date tracking will help you quite a bit, because simply re-editing a line to what it had been (or using Ctrl+Z to undo, if it happened to be the last thing you did) will automatically reset the date to what it had been.

Note that one of the idiosyncrasies of the source date support in this extension is that the date gutter doesn't always update right away. Sometimes you have to do some edit on some other line before you see the new date. So if you think a date should have changed but it hasn't, give it a chance. Go do some other edits and trust that it will update. If you feel the need to see it before moving off to something else, I suggest pressing Enter at the end of the just-edited line. That should create a new blank line, and both of them should have the new date. Then delete the line you just added.

It is similarly disconcerting when you add a line after a blank line, and the "preexisting" blank line has its date updated, but the "newly added" line has the old line's date. This is a side-effect of the diff algorithm, which can't tell which one you just added, only that there are now two blank lines where there had been one, so from a purely logical standpoint, it is equally valid for either one to adopt the new date. Again, you just have to keep using it for a while and get used to how it does things. (And perhaps get into the habit of saving more often, particularly when everything "looks right" to you, which both locks in a "correct" state and reduces the chance that the diff algorithm misinterprets a long series of your edits as a very, very different long series of edits that would have resulted in exactly the same source code.)

If you have watched carefully and think there really has been something amiss, please create an issue on this project to report your findings.

See #819 and #824 for more detailed discussion that went into the development of this feature.
